optimizer/typeinferer.go

```diff
@@ -48,14 +51,111 @@ func (v *typeInferrer) Leave(in ast.Node) (out ast.Node, ok bool) {
 	case *ast.FuncCastExpr:
 		x.SetType(x.Tp)
 	case *ast.SelectStmt:
-		rf := x.GetResultFields()
-		for _, val := range rf {
-			if val.Column.ID == 0 && val.Expr.GetType() != nil {
-				val.Column.FieldType = *(val.Expr.GetType())
+		v.selectStmt(x)
+	case *ast.ParamMarkerExpr:
+		x.SetType(types.DefaultTypeForValue(x.GetValue()))
+	case *ast.BinaryOperationExpr:
+		v.binaryOperation(x)
+	case *ast.UnaryOperationExpr:
+		v.unaryOperation(x)
+	case *ast.BetweenExpr:
+		x.SetType(types.NewFieldType(mysql.TypeLonglong))
+	case *ast.CompareSubqueryExpr:
+		x.SetType(types.NewFieldType(mysql.TypeLonglong))
+	case *ast.ExistsSubqueryExpr:
+		x.SetType(types.NewFieldType(mysql.TypeLonglong))
+	case *ast.PatternInExpr:
+		x.SetType(types.NewFieldType(mysql.TypeLonglong))
+	case *ast.PatternLikeExpr:
+		x.SetType(types.NewFieldType(mysql.TypeLonglong))
+	case *ast.PatternRegexpExpr:
+		x.SetType(types.NewFieldType(mysql.TypeLonglong))
+	case *ast.IsNullExpr:
+		x.SetType(types.NewFieldType(mysql.TypeLonglong))
+	case *ast.IsTruthExpr:
+		x.SetType(types.NewFieldType(mysql.TypeLonglong))
+	case *ast.ParenthesesExpr:
+		x.SetType(x.Expr.GetType())
+		// TODO: handle all expression types.
+	}
+	return in, true
+}
+
+func (v *typeInferrer) selectStmt(x *ast.SelectStmt) {
+	rf := x.GetResultFields()
+	for _, val := range rf {
+		if val.Column.ID == 0 && val.Expr.GetType() != nil {
```

Comment:
`Conlumn.ID = 0` means the column is not a real column from table, but a generated column for result field.